codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/AccountIDEndpointRouting.java

```diff
@@ -0,0 +1,77 @@
+package software.amazon.smithy.aws.go.codegen.customization;
+
+import software.amazon.smithy.aws.go.codegen.SdkGoTypes;
+import software.amazon.smithy.codegen.core.SymbolProvider;
+import software.amazon.smithy.go.codegen.*;
+import software.amazon.smithy.go.codegen.integration.GoIntegration;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.rulesengine.traits.EndpointRuleSetTrait;
+import software.amazon.smithy.utils.MapUtils;
+
+import static software.amazon.smithy.go.codegen.GoWriter.goTemplate;
+
+public class AccountIDEndpointRouting implements GoIntegration {
+    @Override
+    public void renderPreEndpointResolutionHook(GoSettings settings, GoWriter writer, Model model) {
+        writer.write("""
+                if err := checkAccountID(getIdentity(ctx), m.options.AccountIDEndpointMode); err != nil {
+                    return out, metadata, $T("invalid accountID set: %w", err)
+                }
+                """,
+                GoStdlibTypes.Fmt.Errorf);
+    }
+
+    @Override
+    public void writeAdditionalFiles(
+            GoSettings settings,
+            Model model,
+            SymbolProvider symbolProvider,
+            GoDelegator goDelegator
+    ) {
+        if (!settings.getService(model).hasTrait(EndpointRuleSetTrait.class)) {
+            return;
+        }
+        goDelegator.useShapeWriter(settings.getService(model), goTemplate("""
+        func accountID(identity $auth:T, mode $accountIDEndpointMode:T) *string {
```

Comment:
A few things here
1. the name here is too broad, this is likely to collide with random identifiers we generate in the package. I'd call it `resolveAccountID`, `bindAccountID`, etc.
2. you're adding the builtin that calls this in a different location than where it's defined. We should move the `accountID` function you have here to be generated in the builtins integration instead. (checkAccountID _is_ used here and can stay in this integration)